packages/shared/src/authz/domain/loaders.ts

```diff
@@ -8,11 +10,47 @@ import type {
   GetWorkspaceSsoSession
 } from './workspaces/operations.js'
 
-export type AuthCheckContext<LoaderKeys extends keyof AuthCheckContextLoaders> = {
-  loaders: Pick<AuthCheckContextLoaders, LoaderKeys>
+// utility type that ensures all properties functions that return promises
+type PromiseAll<T> = {
+  [K in keyof T]: T[K] extends (...args: infer Args) => MaybeAsync<infer Return>
+    ? (...args: Args) => Promise<Return>
+    : never
 }
 
-export type AuthCheckContextLoaders = {
+// wrapper type for AllAuthCheckContextLoaders that ensures loaders follow the expected schema
+type AuthContextLoaderMappingDefinition<
+  Mapping extends {
+    [Key in keyof Mapping]: Key extends AuthCheckContextLoaderKeys
+      ? Mapping[Key]
+      : never
+  }
+> = PromiseAll<
+  OverrideProperties<
+    {
+      [key in AuthCheckContextLoaderKeys]: unknown
+    },
+    Mapping
+  >
+>
+
+/**
+ * All loaders must be listed here for app startup validation to work properly
+ */
+export const AuthCheckContextLoaderKeys = <const>{
+  getEnv: 'getEnv',
+  getProject: 'getProject',
+  getProjectRole: 'getProjectRole',
+  getServerRole: 'getServerRole',
+  getWorkspace: 'getWorkspace',
+  getWorkspaceRole: 'getWorkspaceRole',
+  getWorkspaceSsoProvider: 'getWorkspaceSsoProvider',
+  getWorkspaceSsoSession: 'getWorkspaceSsoSession'
+}
+
+export type AuthCheckContextLoaderKeys =
+  (typeof AuthCheckContextLoaderKeys)[keyof typeof AuthCheckContextLoaderKeys]
+
+export type AllAuthCheckContextLoaders = AuthContextLoaderMappingDefinition<{
```

Comment:
This wrapping type ensures the loaders follow an expected schema. For example, it forces loaders to be async, even if their return types aren't. Loaders should **always** be async cause they will load things different ways in different environments and there may be caching added ontop